### PR TITLE
Enable authorization directives by default

### LINKED ISSUE #3842: Enable authorization directives by default
Authorization directives should be enabled by default if a valid entitlement is detected. This changes a bit the behaviour of the entitlement check, because it has to check for the presence of the directive in the graph, and must be robust against directive renaming

### PR DESCRIPTION
**Pull request by Geal**
Fix #3842 

This requires a few steps, because authorization directives should not be usable without a valid license, so the license enforcement code has to check for their presence in the graph.

One missing point: making sure that the authorization plugins can recognize when their directives have been renamed (extracted in another PR: https://github.com/apollographql/router/pull/3949)

**Checklist**

Complete the checklist (and note appropriate exceptions) before a final PR is raised.

- [ ] Changes are compatible[^1]
- [ ] Documentation[^2] completed
- [ ] Performance impact assessed and acceptable
- Tests added and passing[^3]
    - [ ] Unit Tests
    - [ ] Integration Tests
    - [ ] Manual Tests

**Exceptions**

*Note any exceptions here*

**Notes**

[^1]. It may be appropriate to bring upcoming changes to the attention of other (impacted) groups. Please endeavour to do this before seeking PR approval. The mechanism for doing this will vary considerably, so use your judgement as to how and when to do this.
[^2]. Configuration is an important part of many changes. Where applicable please try to document configuration examples.
[^3]. Tick whichever testing boxes are applicable. If you are adding Manual Tests:
    - please document the manual testing (extensively) in the Exceptions.
    - please raise a separate issue to automate the test and label it (or ask for it to be labeled) as `manual test`

### CODE REVIEW

File: apollo-router/src/plugins/authorization/scopes.rs [hunk 1]
```diff
@@ -26,23 +26,43 @@ pub(crate) struct ScopeExtractionVisitor<'a> {
...
+                    == Some("https://specs.apollo.dev/requiresScopes/v0.1")
+            })
+            .next()
+        {
+            link.argument_by_name("as")
+                .and_then(|value| value.as_str().map(|s| s.to_string()))
+                .unwrap_or_else(|| REQUIRES_SCOPES_DIRECTIVE_NAME.to_string())
+        } else {
+            return None;
+        };
```
Comment (bnjjj) [inline apollo-router/src/plugins/authorization/scopes.rs:53]: This should be a function to not duplicate that part of code and risk a misspelled name

Comment (bnjjj) [reply apollo-router/src/plugins/authorization/scopes.rs:53]: or create constants for identifiers

File: apollo-router/src/uplink/license_enforcement.rs [hunk 2]
```diff
@@ -183,17 +216,47 @@ impl LicenseEnforcementReport {
                 .build(),
         ]
     }
+
+    fn schema_restrictions() -> Vec<SchemaRestriction> {
+        vec![
+            SchemaRestriction::builder()
+                .name("@authenticated")
+                .url("https://specs.apollo.dev/authenticated/v0.1")
```
Comment (bnjjj) [inline apollo-router/src/uplink/license_enforcement.rs:224]: const ?

File: apollo-router/src/uplink/license_enforcement.rs [hunk 3]
```diff
@@ -183,17 +216,47 @@ impl LicenseEnforcementReport {
...
+                .build(),
+            SchemaRestriction::builder()
+                .name("@requiresScopes")
+                .url("https://specs.apollo.dev/requiresScopes/v0.1")
```
Comment (bnjjj) [inline apollo-router/src/uplink/license_enforcement.rs:228]: const ?

File: apollo-router/src/uplink/license_enforcement.rs [hunk 4]
```diff
@@ -119,6 +127,31 @@ impl LicenseEnforcementReport {
...
     }
 
+    fn validate_schema(
+        schema: &ApolloCompiler,
+        schema_restrictions: &Vec<SchemaRestriction>,
+    ) -> Vec<SchemaRestriction> {
+        let feature_urls = schema
+            .db
+            .schema()
+            .directives_by_name("link")
```
[the remark this answers is not in the document]
Comment (Geal) [reply apollo-router/src/uplink/license_enforcement.rs]: fixed in [04962fe](https://github.com/apollographql/router/pull/3713/commits/04962fe264a2e4b67dcf2546a222c1e709443a4c)

File: apollo-router/src/plugins/authorization/scopes.rs (hunk 1 above)
Comment (Geal) [reply apollo-router/src/plugins/authorization/scopes.rs]: moved to https://github.com/apollographql/router/pull/3949

File: apollo-router/src/uplink/license_enforcement.rs (hunk 2 above)
Comment (Geal) [reply apollo-router/src/uplink/license_enforcement.rs:226]: moved to https://github.com/apollographql/router/pull/3949

File: apollo-router/src/uplink/license_enforcement.rs (hunk 3 above)
Comment (Geal) [reply apollo-router/src/uplink/license_enforcement.rs:230]: moved to https://github.com/apollographql/router/pull/3949

**Review by bnjjj — APPROVED** [review]

File: .changesets/feat_geal_authz_optout.md [hunk 5]
```diff
@@ -0,0 +1,5 @@
+### Authorization directives are enabled by default ([Issue #3842](https://github.com/apollographql/router/issues/3842))
+
+If the router starts with an API key from an Enterprise account, and the schema contains the authorization directives, then they will be usable directly wxithout further configuration.
```
Comment (o0Ignition0o) [inline .changesets/feat_geal_authz_optout.md:3]:
```suggestion
If the router starts with an API key from an Enterprise account, and the schema contains the authorization directives, then they will be usable directly xithout further configuration.
```

**Review by o0Ignition0o — APPROVED** [review]

**Pull request merged**